src/mlpack/methods/ann/layer/grouped_convolution_impl.hpp

```diff
@@ -0,0 +1,686 @@
+    GradientConvolutionRule,
+    MatType
+>&
+GroupedConvolutionType<
+    ForwardConvolutionRule,
+    BackwardConvolutionRule,
+    GradientConvolutionRule,
+    MatType
+>::operator=(GroupedConvolutionType&& other)
+{
+  if (&other != this)
+  {
+    Layer<MatType>::operator=(std::move(other));
+    maps = std::move(other.maps);
+    kernelWidth = std::move(other.kernelWidth);
+    kernelHeight = std::move(other.kernelHeight);
+    groups = std::move(other.groups);
+    strideWidth = std::move(other.strideWidth);
+    strideHeight = std::move(other.strideHeight);
+    padWLeft = std::move(other.padWLeft);
+    padWRight = std::move(other.padWRight);
+    padHBottom = std::move(other.padHBottom);
+    padHTop = std::move(other.padHTop);
+    useBias = std::move(other.useBias);
+    padding = std::move(other.padding);
+    paddingType = std::move(other.paddingType);
+    inMaps = std::move(other.inMaps);
+    higherInDimensions = std::move(other.higherInDimensions);
+  }
+
+  return *this;
+}
+
+template<
+    typename ForwardConvolutionRule,
+    typename BackwardConvolutionRule,
+    typename GradientConvolutionRule,
+    typename MatType
+>
+void GroupedConvolutionType<
+    ForwardConvolutionRule,
+    BackwardConvolutionRule,
+    GradientConvolutionRule,
+    MatType
+>::SetWeights(typename MatType::elem_type* weightPtr)
+{
+  MakeAlias(weight, weightPtr, kernelWidth, kernelHeight, 
+      (maps * inMaps) / groups);
+    if (useBias)
+  {
+    MakeAlias(bias, weightPtr + weight.n_elem, maps, 1);
+    MakeAlias(weights, weightPtr, weight.n_elem + bias.n_elem, 1);
+  }
+  else
+  {
+    MakeAlias(weights, weightPtr, weight.n_elem, 1);
+  }
+}
+
+template<
+    typename ForwardConvolutionRule,
+    typename BackwardConvolutionRule,
+    typename GradientConvolutionRule,
+    typename MatType
+>
+void GroupedConvolutionType<
+    ForwardConvolutionRule,
+    BackwardConvolutionRule,
+    GradientConvolutionRule,
+    MatType
+>::Forward(const MatType& input, MatType& output)
+{
+  batchSize = input.n_cols;
+
+  // First, perform any padding if necessary.
+  const bool usingPadding =
+      (padWLeft != 0 || padWRight != 0 || padHTop != 0 || padHBottom != 0);
+  const size_t paddedRows = this->inputDimensions[0] + padWLeft + padWRight;
+  const size_t paddedCols = this->inputDimensions[1] + padHTop + padHBottom;
+  if (usingPadding)
+  {
+    inputPadded.set_size(paddedRows * paddedCols * inMaps * higherInDimensions,
+        input.n_cols);
+    padding.Forward(input, inputPadded);
+  }
+
+  arma::Cube<typename MatType::elem_type> inputTemp;
+  MakeAlias(inputTemp,
+      const_cast<MatType&>(usingPadding ? inputPadded : input).memptr(),
+      paddedRows, paddedCols, inMaps * higherInDimensions * batchSize);
+
+  MakeAlias(outputTemp, output.memptr(), this->outputDimensions[0],
+      this->outputDimensions[1], maps * higherInDimensions * batchSize);
+  outputTemp.zeros();
+
+  size_t inGroupSize = inMaps / groups;
```

Comment:
What happens if `inMaps` is not divisible by `groups`?  e.g., if `groups` is `3` but `inMaps` is `7`, then we get `inGroupSize` as `2` (since integer division will effectively floor any result), and the seventh input mapping that the user asked for gets ignored.  Should we instead take the ceiling of the integer division, and then have the last group potentially be smaller than the first groups?  In my view that would be how I would expect this layer to behave, but I'm open to other possibilities too, so long as it is consistent behavior. :+1:

Reply:
You can see `ComputeOutputDimensions()`. It will throw an error there.

Reply:
Ahh, thank you for pointing that out.